fuse_optimizers/src/fixed_lag_smoother.cpp

```diff
@@ -0,0 +1,320 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2019, Locus Robotics
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <fuse_optimizers/fixed_lag_smoother.h>
+
+#include <fuse_constraints/marginalize_variables.h>
+#include <fuse_core/graph.h>
+#include <fuse_core/transaction.h>
+#include <fuse_core/uuid.h>
+#include <fuse_optimizers/optimizer.h>
+#include <ros/ros.h>
+
+#include <algorithm>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <vector>
+
+
+namespace fuse_optimizers
+{
+
+FixedLagSmoother::FixedLagSmoother(
+  fuse_core::Graph::UniquePtr graph,
+  const ros::NodeHandle& node_handle,
+  const ros::NodeHandle& private_node_handle) :
+    fuse_optimizers::Optimizer(std::move(graph), node_handle, private_node_handle),
+    optimization_request_(false),
+    optimization_running_(true),
+    start_time_(ros::TIME_MAX),
+    started_(false)
+{
+  double lag_duration;
+  double default_lag_duration = 5.0;
+  private_node_handle_.param("lag_duration", lag_duration, default_lag_duration);
+  if (lag_duration <= 0)
+  {
+    ROS_WARN_STREAM("The requested lag_duration is <= 0. Using the default value (" <<
+                    default_lag_duration << "s) instead.");
+    lag_duration = default_lag_duration;
+  }
+  lag_duration_.fromSec(lag_duration);
+
+  double optimization_frequency;
```

Comment:
We repeat this identical pattern a bunch of times. Maybe make a `getPositiveParam` method or something?